### PR TITLE
PV usage metrics to ETL

### PR DESCRIPTION
**Pull request by nickcurie**
## What does this PR change?
* Pulls in PV usage metrics into ETL

## Does this PR relate to any other PRs?
* [KCM#963](https://github.com/kubecost/kubecost-cost-model/pull/963)
* [helm-chart#1713](https://github.com/kubecost/cost-analyzer-helm-chart/pull/1713)

## How will this PR impact users?
* Users can see average byte usage and max byte usage for PVs

## Does this PR address any GitHub or Zendesk issues?
* 

## How was this PR tested?
* Locally

## Does this PR require changes to documentation?
* Yes

## Have you labeled this PR and its corresponding Issue as "next release" if it should be part of the next Opencost release? If not, why not?
* v1.98

### CODE REVIEW

File: pkg/costmodel/cluster.go [hunk 1]
```diff
@@ -1291,4 +1341,120 @@ func pvCosts(diskMap map[DiskIdentifier]*Disk, resolution time.Duration, resActi
...
+		claimNamespace, err := result.GetString("namespace")
+		if err != nil {
+			log.Warnf("ClusterDisks: pv usage data missing namespace")
+			continue
+		}
+
+		var volumeName string
+
+		for _, thatRes := range resPVCInfo {
+
```
Comment (michaelmdresser) [inline pkg/costmodel/cluster.go:1421]: We glossed over the inner loop here while working on functionality. Now that we've got functionality working, do you see a way to avoid this inner loop? Maybe by constructing an intermediate `map` that gives you the data mapping you need?

File: pkg/costmodel/cluster.go [hunk 2]
```diff
@@ -1291,4 +1341,120 @@ func pvCosts(diskMap map[DiskIdentifier]*Disk, resolution time.Duration, resActi
...
+				log.Warnf("ClusterDisks: pv claim data missing volumename")
+			}
+			thatClaimName, err := thatRes.GetString("persistentvolumeclaim")
+			if err != nil {
+				log.Warnf("ClusterDisks: pv claim data missing persistentvolumeclaim")
+			}
+			thatClaimNamespace, err := thatRes.GetString("namespace")
+			if err != nil {
+				log.Warnf("ClusterDisks: pv claim data missing namespace")
+			}
```
Comment (michaelmdresser) [inline pkg/costmodel/cluster.go]: In these error cases, should you have a `continue` after the `log.Warnf`?

File: pkg/costmodel/cluster.go [hunk 3]
```diff
@@ -1291,4 +1341,120 @@ func pvCosts(diskMap map[DiskIdentifier]*Disk, resolution time.Duration, resActi
...
+		claimNamespace, err := result.GetString("namespace")
+		if err != nil {
+			log.Warnf("ClusterDisks: pv usage data missing namespace")
+			continue
+		}
+
+		var volumeName string
+
+		for _, thatRes := range resPVCInfo {
+
```
Comment (michaelmdresser) [inline pkg/costmodel/cluster.go:1482]: Same q

File: pkg/costmodel/cluster.go [hunk 4]
```diff
@@ -1291,4 +1341,120 @@ func pvCosts(diskMap map[DiskIdentifier]*Disk, resolution time.Duration, resActi
...
+				log.Warnf("ClusterDisks: pv claim data missing volumename")
+			}
+			thatClaimName, err := thatRes.GetString("persistentvolumeclaim")
+			if err != nil {
+				log.Warnf("ClusterDisks: pv claim data missing persistentvolumeclaim")
+			}
+			thatClaimNamespace, err := thatRes.GetString("namespace")
+			if err != nil {
+				log.Warnf("ClusterDisks: pv claim data missing namespace")
+			}
```
Comment (michaelmdresser) [inline pkg/costmodel/cluster.go]: Same q

File: pkg/costmodel/cluster.go [hunk 5]
```diff
@@ -107,17 +107,22 @@ func NewClusterCostsFromCumulative(cpu, gpu, ram, storage float64, window, offse
...
+	Name           string
+	ProviderID     string
+	StorageClass   string
+	VolumeName     string
+	ClaimName      string
+	ClaimNamespace string
+	Cost           float64
+	Bytes          float64
+	BytesUsedAvg   float64
+	BytesUsedMax   float64
```
Comment (michaelmdresser) [inline pkg/costmodel/cluster.go:120]: I think you had a question about these because they aren't getting filled for local disks, just for PVs. Is that still right? What approaches have you considered?

File: pkg/kubecost/asset_json.go [hunk 6]
```diff
@@ -332,6 +338,17 @@ func (d *Disk) InterfaceToDisk(itf interface{}) error {
 	if ByteHours, err := getTypedVal(fmap["byteHours"]); err == nil {
 		d.ByteHours = ByteHours.(float64)
 	}
+	if ByteHoursUsed, err := getTypedVal(fmap["byteHoursUsed"]); err == nil {
+		d.ByteHoursUsed = ByteHoursUsed.(float64)
+	}
+	if ByteUsageMax, err := getTypedVal(fmap["byteUsageMax"]); err == nil {
+		if ByteUsageMax == nil {
+			d.ByteUsageMax = nil
```
Comment (michaelmdresser) [inline pkg/kubecost/asset_json.go:349]: Would you add a case to `TestDisk_Unmarshal` that tests this null unmarshal?

File: pkg/kubecost/bingen.go [hunk 7]
```diff
@@ -71,4 +71,4 @@ package kubecost
 // @bingen:generate:AuditSetRange
 // @bingen:end
 
-//go:generate bingen -package=kubecost -version=15 -buffer=github.com/opencost/opencost/pkg/util
+//go:generate bingen -package=kubecost -version=18 -buffer=github.com/opencost/opencost/pkg/util
```
Comment (michaelmdresser) [inline pkg/kubecost/bingen.go]: I'm pretty sure you don't have to change the version here -- this only changes the version for things that don't specify their own version, like `Window`.

File: pkg/costmodel/cluster.go (hunk 5 above)
Comment (nickcurie) [reply pkg/costmodel/cluster.go:120]: We have metrics in prom that track container file system usage. I've included these prom queries in the latest commit, so local disks have avg and max usage in ETL.

File: pkg/costmodel/cluster.go (hunk 1 above)
Comment (nickcurie) [reply pkg/costmodel/cluster.go:1421]: I think its worth coming back to this, since this is an efficiency ask. I think its valuable to get this PR out so that users can begin to interact with it.

**Review by mbolt35 — APPROVED** [review]
Looks good to me! Great work
